File: tests/algorithms/x/Java/matrix/nth_fibonacci_using_matrix_exponentiation.java
public class Main {

    static int[][] multiply(int[][] matrix_a, int[][] matrix_b) {
        int n = matrix_a.length;
        int[][] matrix_c = ((int[][])(new int[][]{}));
        int i = 0;
        while (i < n) {
            int[] row = ((int[])(new int[]{}));
            int j = 0;
            while (j < n) {
                int val = 0;
                int k = 0;
                while (k < n) {
                    val = val + matrix_a[i][k] * matrix_b[k][j];
                    k = k + 1;
                }
                row = ((int[])(java.util.stream.IntStream.concat(java.util.Arrays.stream(row), java.util.stream.IntStream.of(val)).toArray()));
                j = j + 1;
            }
            matrix_c = ((int[][])(appendObj(matrix_c, row)));
            i = i + 1;
        }
        return matrix_c;
    }

    static int[][] identity(int n) {
        int[][] res = ((int[][])(new int[][]{}));
        int i_1 = 0;
        while (i_1 < n) {
            int[] row_1 = ((int[])(new int[]{}));
            int j_1 = 0;
            while (j_1 < n) {
                if (i_1 == j_1) {
                    row_1 = ((int[])(java.util.stream.IntStream.concat(java.util.Arrays.stream(row_1), java.util.stream.IntStream.of(1)).toArray()));
                } else {
                    row_1 = ((int[])(java.util.stream.IntStream.concat(java.util.Arrays.stream(row_1), java.util.stream.IntStream.of(0)).toArray()));
                }
                j_1 = j_1 + 1;
            }
            res = ((int[][])(appendObj(res, row_1)));
            i_1 = i_1 + 1;
        }
        return res;
    }

    static int nth_fibonacci_matrix(int n) {
        if (n <= 1) {
            return n;
        }
        int[][] res_matrix = ((int[][])(identity(2)));
        int[][] fib_matrix = ((int[][])(new int[][]{new int[]{1, 1}, new int[]{1, 0}}));
        int m = n - 1;
        while (m > 0) {
            if (Math.floorMod(m, 2) == 1) {
                res_matrix = ((int[][])(multiply(((int[][])(res_matrix)), ((int[][])(fib_matrix)))));
            }
            fib_matrix = ((int[][])(multiply(((int[][])(fib_matrix)), ((int[][])(fib_matrix)))));
            m = Math.floorDiv(m, 2);
        }
        return res_matrix[0][0];
    }

    static int nth_fibonacci_bruteforce(int n) {
        if (n <= 1) {
            return n;
        }
        int fib0 = 0;
        int fib1 = 1;
        int i_2 = 2;
        while (i_2 <= n) {
            int next = fib0 + fib1;
            fib0 = fib1;
            fib1 = next;
            i_2 = i_2 + 1;
        }
        return fib1;
    }

    static int parse_number(String s) {
        int result = 0;
        int i_3 = 0;
        while (i_3 < _runeLen(s)) {
            String ch = _substr(s, i_3, i_3 + 1);
            if ((ch.compareTo("0") >= 0) && (ch.compareTo("9") <= 0)) {
                result = result * 10 + (Integer.parseInt(ch));
            }
            i_3 = i_3 + 1;
        }
        return result;
    }

    static void main() {
        String[] ordinals = ((String[])(new String[]{"0th", "1st", "2nd", "3rd", "10th", "100th", "1000th"}));
        int i_4 = 0;
        while (i_4 < ordinals.length) {
            String ordinal = ordinals[i_4];
            int n_1 = parse_number(ordinal);
            String msg = ordinal + " fibonacci number using matrix exponentiation is " + _p(nth_fibonacci_matrix(n_1)) + " and using bruteforce is " + _p(nth_fibonacci_bruteforce(n_1));
            System.out.println(msg);
            i_4 = i_4 + 1;
        }
    }
    public static void main(String[] args) {
        {
            long _benchStart = _now();
            long _benchMem = _mem();
            main();
            long _benchDuration = _now() - _benchStart;
            long _benchMemory = _mem() - _benchMem;
            System.out.println("{");
            System.out.println("  \"duration_us\": " + _benchDuration + ",");
            System.out.println("  \"memory_bytes\": " + _benchMemory + ",");
            System.out.println("  \"name\": \"main\"");
            System.out.println("}");
            return;
        }
    }

    static boolean _nowSeeded = false;
    static int _nowSeed;
    static int _now() {
        if (!_nowSeeded) {
            String s = System.getenv("MOCHI_NOW_SEED");
            if (s != null && !s.isEmpty()) {
                try { _nowSeed = Integer.parseInt(s); _nowSeeded = true; } catch (Exception e) {}
            }
        }
        if (_nowSeeded) {
            _nowSeed = (int)((_nowSeed * 1664525L + 1013904223) % 2147483647);
            return _nowSeed;
        }
        return (int)(System.nanoTime() / 1000);
    }

    static long _mem() {
        Runtime rt = Runtime.getRuntime();
        rt.gc();
        return rt.totalMemory() - rt.freeMemory();
    }

    static <T> T[] appendObj(T[] arr, T v) {
        T[] out = java.util.Arrays.copyOf(arr, arr.length + 1);
        out[arr.length] = v;
        return out;
    }

    static int _runeLen(String s) {
        return s.codePointCount(0, s.length());
    }

    static String _substr(String s, int i, int j) {
        int start = s.offsetByCodePoints(0, i);
        int end = s.offsetByCodePoints(0, j);
        return s.substring(start, end);
    }

    static String _p(Object v) {
        if (v == null) return "<nil>";
        if (v.getClass().isArray()) {
            if (v instanceof int[]) return java.util.Arrays.toString((int[]) v);
            if (v instanceof long[]) return java.util.Arrays.toString((long[]) v);
            if (v instanceof double[]) return java.util.Arrays.toString((double[]) v);
            if (v instanceof boolean[]) return java.util.Arrays.toString((boolean[]) v);
            if (v instanceof byte[]) return java.util.Arrays.toString((byte[]) v);
            if (v instanceof char[]) return java.util.Arrays.toString((char[]) v);
            if (v instanceof short[]) return java.util.Arrays.toString((short[]) v);
            if (v instanceof float[]) return java.util.Arrays.toString((float[]) v);
            return java.util.Arrays.deepToString((Object[]) v);
        }
        return String.valueOf(v);
    }
}

File: tests/algorithms/x/PHP/matrix/nth_fibonacci_using_matrix_exponentiation.php
<?php
ini_set('memory_limit', '-1');
$now_seed = 0;
$now_seeded = false;
$s = getenv('MOCHI_NOW_SEED');
if ($s !== false && $s !== '') {
    $now_seed = intval($s);
    $now_seeded = true;
}
function _now() {
    global $now_seed, $now_seeded;
    if ($now_seeded) {
        $now_seed = ($now_seed * 1664525 + 1013904223) % 2147483647;
        return $now_seed;
    }
    return hrtime(true);
}
function _str($x) {
    if (is_array($x)) {
        $isList = array_keys($x) === range(0, count($x) - 1);
        if ($isList) {
            $parts = [];
            foreach ($x as $v) { $parts[] = _str($v); }
            return '[' . implode(' ', $parts) . ']';
        }
        $parts = [];
        foreach ($x as $k => $v) { $parts[] = _str($k) . ':' . _str($v); }
        return 'map[' . implode(' ', $parts) . ']';
    }
    if (is_bool($x)) return $x ? 'true' : 'false';
    if ($x === null) return 'null';
    return strval($x);
}
function _append($arr, $x) {
    $arr[] = $x;
    return $arr;
}
function _intdiv($a, $b) {
    if (function_exists('bcdiv')) {
        $sa = is_int($a) ? strval($a) : (is_string($a) ? $a : sprintf('%.0f', $a));
        $sb = is_int($b) ? strval($b) : (is_string($b) ? $b : sprintf('%.0f', $b));
        return intval(bcdiv($sa, $sb, 0));
    }
    return intdiv($a, $b);
}
$__start_mem = memory_get_usage();
$__start = _now();
  function multiply($matrix_a, $matrix_b) {
  $n = count($matrix_a);
  $matrix_c = [];
  $i = 0;
  while ($i < $n) {
  $row = [];
  $j = 0;
  while ($j < $n) {
  $val = 0;
  $k = 0;
  while ($k < $n) {
  $val = $val + $matrix_a[$i][$k] * $matrix_b[$k][$j];
  $k = $k + 1;
};
  $row = _append($row, $val);
  $j = $j + 1;
};
  $matrix_c = _append($matrix_c, $row);
  $i = $i + 1;
};
  return $matrix_c;
};
  function identity($n) {
  $res = [];
  $i = 0;
  while ($i < $n) {
  $row = [];
  $j = 0;
  while ($j < $n) {
  if ($i == $j) {
  $row = _append($row, 1);
} else {
  $row = _append($row, 0);
}
  $j = $j + 1;
};
  $res = _append($res, $row);
  $i = $i + 1;
};
  return $res;
};
  function nth_fibonacci_matrix($n) {
  if ($n <= 1) {
  return $n;
}
  $res_matrix = identity(2);
  $fib_matrix = [[1, 1], [1, 0]];
  $m = $n - 1;
  while ($m > 0) {
  if ($m % 2 == 1) {
  $res_matrix = multiply($res_matrix, $fib_matrix);
}
  $fib_matrix = multiply($fib_matrix, $fib_matrix);
  $m = _intdiv($m, 2);
};
  return $res_matrix[0][0];
};
  function nth_fibonacci_bruteforce($n) {
  if ($n <= 1) {
  return $n;
}
  $fib0 = 0;
  $fib1 = 1;
  $i = 2;
  while ($i <= $n) {
  $next = $fib0 + $fib1;
  $fib0 = $fib1;
  $fib1 = $next;
  $i = $i + 1;
};
  return $fib1;
};
  function parse_number($s) {
  $result = 0;
  $i = 0;
  while ($i < strlen($s)) {
  $ch = substr($s, $i, $i + 1 - $i);
  if ($ch >= '0' && $ch <= '9') {
  $result = $result * 10 + (intval($ch));
}
  $i = $i + 1;
};
  return $result;
};
  function main() {
  $ordinals = ['0th', '1st', '2nd', '3rd', '10th', '100th', '1000th'];
  $i = 0;
  while ($i < count($ordinals)) {
  $ordinal = $ordinals[$i];
  $n = parse_number($ordinal);
  $msg = $ordinal . ' fibonacci number using matrix exponentiation is ' . _str(nth_fibonacci_matrix($n)) . ' and using bruteforce is ' . _str(nth_fibonacci_bruteforce($n));
  echo rtrim($msg), PHP_EOL;
  $i = $i + 1;
};
};
  main();
$__end = _now();
$__end_mem = memory_get_peak_usage();
$__duration = max(1, intdiv($__end - $__start, 1000));
$__mem_diff = max(0, $__end_mem - $__start_mem);
$__bench = ["duration_us" => $__duration, "memory_bytes" => $__mem_diff, "name" => "main"];
$__j = json_encode($__bench, 128);
$__j = str_replace("    ", "  ", $__j);
echo $__j, PHP_EOL;
